File: SOTALog/SOTALog/Services/ADIFFormatter.swift
import Foundation
import CoreTransferable
import UniformTypeIdentifiers

/// Encodes and decodes ADIF (Amateur Data Interchange Format) records.
enum ADIFFormatter {

    /// Program-specific export filtering.
    enum Program {
        case pota
        case sota
    }

    // MARK: - Encoding

    /// Encodes a single QSO to an ADIF record string.
    static func encode(qso: QSO, log: Log? = nil) -> String {
        encode(qso: qso, log: log, program: nil)
    }

    /// Encodes a single QSO, filtering fields for a specific program.
    static func encode(qso: QSO, log: Log? = nil, program: Program?) -> String {
        var fields: [(String, String)] = []

        fields.append(("CALL", qso.callsign))
        fields.append(("QSO_DATE", qso.date))
        fields.append(("TIME_ON", qso.timeOn))
        fields.append(("BAND", qso.band))
        fields.append(("MODE", qso.mode))
        fields.append(("RST_SENT", qso.rstSent))
        fields.append(("RST_RCVD", qso.rstReceived))

        if let freq = qso.frequency {
            fields.append(("FREQ", String(format: "%.4f", freq)))
        }
        if let name = qso.name, !name.isEmpty {
            fields.append(("NAME", name))
        }
        if let qth = qso.qth, !qth.isEmpty {
            fields.append(("QTH", qth))
        }
        if let grid = qso.grid, !grid.isEmpty {
            fields.append(("GRIDSQUARE", grid))
        }
        if let notes = qso.notes, !notes.isEmpty {
            fields.append(("COMMENT", notes))
        }

        if let log = log {
            // POTA fields — skip for SOTA exports
            if program != .sota, let myRef = log.potaReference {
                fields.append(("MY_SIG", "POTA"))
                fields.append(("MY_SIG_INFO", myRef))
            }
            if let myGrid = log.myGrid {
                fields.append(("MY_GRIDSQUARE", myGrid))
            }
            fields.append(("STATION_CALLSIGN", log.myCallsign))

            // SOTA fields — skip for POTA exports
            if program != .pota, let myRef = log.sotaReference {
                fields.append(("MY_SOTA_REF", myRef))
            }
        }

        // Their POTA ref (park-to-park) — skip for SOTA exports
        if program != .sota, let ref = qso.potaRef, !ref.isEmpty {
            fields.append(("SIG", "POTA"))
            fields.append(("SIG_INFO", ref))
        }

        // Their SOTA ref (summit-to-summit) — skip for POTA exports
        if program != .pota, let ref = qso.sotaRef, !ref.isEmpty {
            fields.append(("SOTA_REF", ref))
        }

        let record = fields.map { encodeField($0.0, value: $0.1) }.joined()
        return record + "<EOR>\n"
    }

    /// Encodes a full ADIF file with header.
    static func encodeFile(qsos: [QSO], log: Log? = nil) -> String {
        encodeFile(qsos: qsos, log: log, program: nil)
    }

    /// Encodes a full ADIF file, filtering fields for a specific program.
    static func encodeFile(qsos: [QSO], log: Log? = nil, program: Program?) -> String {
        var output = "ADIF Export from SOTALog\n"
        output += encodeField("ADIF_VER", value: "3.1.4")
        output += encodeField("PROGRAMID", value: "SOTALog")
        output += encodeField("PROGRAMVERSION", value: "1.0")
        output += "<EOH>\n\n"

        for qso in qsos {
            output += encode(qso: qso, log: log, program: program)
        }

        return output
    }

    /// Encodes a full ADIF file from log+QSO sections, preserving each log's context.
    static func encodeFile(sections: [(Log, [QSO])]) -> String {
        var output = "ADIF Export from SOTALog\n"
        output += encodeField("ADIF_VER", value: "3.1.4")
        output += encodeField("PROGRAMID", value: "SOTALog")
        output += encodeField("PROGRAMVERSION", value: "1.0")
        output += "<EOH>\n\n"

        for (log, qsos) in sections {
            for qso in qsos {
                output += encode(qso: qso, log: log)
            }
        }

        return output
    }

    private static func encodeField(_ name: String, value: String) -> String {
        "<\(name):\(value.count)>\(value)"
    }

    // MARK: - Decoding

    /// Parses ADIF text into an array of field dictionaries.
    static func decode(_ adif: String) -> [[String: String]] {
        var records: [[String: String]] = []

        // Skip header if present — find <EOH> or start from beginning
        let bodyStart: String.Index
        if let eohRange = adif.range(of: "<EOH>", options: .caseInsensitive) {
            bodyStart = eohRange.upperBound
        } else if let eohRange = adif.range(of: "<eoh>", options: .caseInsensitive) {
            bodyStart = eohRange.upperBound
        } else {
            bodyStart = adif.startIndex
        }

        let body = String(adif[bodyStart...])

        // Normalize <eor>/<Eor>/etc. to <EOR>, then split
        let normalized = body.replacingOccurrences(of: "<eor>", with: "<EOR>", options: .caseInsensitive)
        let rawRecords = normalized.components(separatedBy: "<EOR>")

        for rawRecord in rawRecords {
            let fields = parseFields(rawRecord)
            if !fields.isEmpty {
                records.append(fields)
            }
        }

        return records
    }

    /// Parses a single record's fields.
    private static func parseFields(_ record: String) -> [String: String] {
        var fields: [String: String] = [:]
        var index = record.startIndex

        while index < record.endIndex {
            // Find next '<'
            guard let openBracket = record[index...].firstIndex(of: "<") else { break }
            guard let closeBracket = record[openBracket...].firstIndex(of: ">") else { break }

            let tagContent = String(record[record.index(after: openBracket)..<closeBracket])
            let parts = tagContent.split(separator: ":", maxSplits: 2)

            guard parts.count >= 2, let length = Int(parts[1]) else {
                index = record.index(after: closeBracket)
                continue
            }

            let fieldName = String(parts[0]).uppercased()
            let valueStart = record.index(after: closeBracket)

            guard valueStart < record.endIndex else {
                index = valueStart
                continue
            }

            let valueEnd = record.index(valueStart, offsetBy: length, limitedBy: record.endIndex) ?? record.endIndex
            let value = String(record[valueStart..<valueEnd])

            fields[fieldName] = value
            index = valueEnd
        }

        return fields
    }

    // MARK: - Filenames

    /// Generates an export filename for the given log and program.
    static func filename(log: Log, program: Program?) -> String {
        switch program {
        case .pota:
            return "\(log.myCallsign)@\(log.potaReference ?? "POTA")_\(log.date).adi"
        case .sota:
            if let ref = log.sotaReference {
                return "\(log.myCallsign)@\(ref.replacingOccurrences(of: "/", with: "-"))_\(log.date).adi"
            }
            return "\(log.myCallsign)_SOTA_\(log.date).adi"
        case nil:
            return "\(log.myCallsign)_\(log.date).adi"
        }
    }

    /// Converts parsed ADIF fields into a QSO record.
    static func qsoFromFields(_ fields: [String: String], logId: Int64? = nil) -> QSO? {
        guard let callsign = fields["CALL"],
              let date = fields["QSO_DATE"],
              let timeOn = fields["TIME_ON"] else {
            return nil
        }

        let band: String = {
            if let freqStr = fields["FREQ"], let freq = Double(freqStr),
               let derived = BandPlan.band(for: freq) {
                return derived
            }
            if let raw = fields["BAND"], BandPlan.allBands.contains(raw) {
                return raw
            }
            return "20m"
        }()

        return QSO(
            logId: logId,
            callsign: callsign.uppercased(),
            date: date,
            timeOn: String(timeOn.prefix(4)),
            frequency: fields["FREQ"].flatMap(Double.init),
            band: band,
            mode: fields["MODE"] ?? "CW",
            rstSent: fields["RST_SENT"] ?? "599",
            rstReceived: fields["RST_RCVD"] ?? "599",
            name: fields["NAME"],
            qth: fields["QTH"],
            grid: fields["GRIDSQUARE"],
            sotaRef: fields["SOTA_REF"],
            potaRef: fields["SIG_INFO"],
            notes: fields["COMMENT"]
        )
    }
}

/// A named ADIF file that can be shared via ShareLink.
struct ADIFFile: Transferable {
    let filename: String
    let content: String

    static var transferRepresentation: some TransferRepresentation {
        FileRepresentation(exportedContentType: .plainText) { file in
            let url = FileManager.default.temporaryDirectory
                .appendingPathComponent(file.filename)
            try file.content.write(to: url, atomically: true, encoding: .utf8)
            return SentTransferredFile(url)
        }
    }
}

File: android/app/src/main/kotlin/com/sotalog/android/domain/services/ADIFFormatter.kt
package com.sotalog.android.domain.services

import com.sotalog.android.domain.models.Log
import com.sotalog.android.domain.models.QSO
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

object ADIFFormatter {

    enum class Program { POTA, SOTA }

    // MARK: - Encoding

    fun encode(qso: QSO, log: Log? = null, program: Program? = null): String {
        val fields = mutableListOf<Pair<String, String>>()

        fields += "CALL" to qso.callsign
        fields += "QSO_DATE" to qso.date
        fields += "TIME_ON" to qso.timeOn
        fields += "BAND" to qso.band
        fields += "MODE" to qso.mode
        fields += "RST_SENT" to qso.rstSent
        fields += "RST_RCVD" to qso.rstReceived

        qso.frequency?.let { fields += "FREQ" to "%.4f".format(it) }
        qso.name?.takeIf { it.isNotEmpty() }?.let { fields += "NAME" to it }
        qso.qth?.takeIf { it.isNotEmpty() }?.let { fields += "QTH" to it }
        qso.grid?.takeIf { it.isNotEmpty() }?.let { fields += "GRIDSQUARE" to it }
        qso.notes?.takeIf { it.isNotEmpty() }?.let { fields += "COMMENT" to it }

        if (log != null) {
            if (program != Program.SOTA) {
                log.potaReference?.let {
                    fields += "MY_SIG" to "POTA"
                    fields += "MY_SIG_INFO" to it
                }
            }
            log.myGrid?.let { fields += "MY_GRIDSQUARE" to it }
            fields += "STATION_CALLSIGN" to log.myCallsign

            if (program != Program.POTA) {
                log.sotaReference?.let { fields += "MY_SOTA_REF" to it }
            }
        }

        if (program != Program.SOTA) {
            qso.potaRef?.takeIf { it.isNotEmpty() }?.let {
                fields += "SIG" to "POTA"
                fields += "SIG_INFO" to it
            }
        }

        if (program != Program.POTA) {
            qso.sotaRef?.takeIf { it.isNotEmpty() }?.let {
                fields += "SOTA_REF" to it
            }
        }

        val record = fields.joinToString("") { encodeField(it.first, it.second) }
        return "$record<EOR>\n"
    }

    fun encodeFile(
        qsos: List<QSO>,
        log: Log? = null,
        program: Program? = null,
    ): String = buildString {
        append("ADIF Export from SOTA Log\n")
        append(encodeField("ADIF_VER", "3.1.4"))
        append(encodeField("PROGRAMID", "SOTA Log"))
        append(encodeField("PROGRAMVERSION", "1.0"))
        append("<EOH>\n\n")
        for (qso in qsos) {
            append(encode(qso, log, program))
        }
    }

    fun encodeFile(
        sections: List<Pair<Log, List<QSO>>>,
        unattached: List<QSO> = emptyList(),
    ): String = buildString {
        append("ADIF Export from SOTA Log\n")
        append(encodeField("ADIF_VER", "3.1.4"))
        append(encodeField("PROGRAMID", "SOTA Log"))
        append(encodeField("PROGRAMVERSION", "1.0"))
        append("<EOH>\n\n")
        for ((log, qsos) in sections) {
            for (qso in qsos) {
                append(encode(qso, log))
            }
        }
        for (qso in unattached) {
            append(encode(qso))
        }
    }

    fun encodeField(name: String, value: String): String =
        "<$name:${value.length}>$value"

    // MARK: - Decoding

    fun decode(adif: String): List<Map<String, String>> {
        val eohIndex = adif.indexOf("<EOH>", ignoreCase = true)
        val body = if (eohIndex >= 0) {
            adif.substring(eohIndex + 5)
        } else {
            adif
        }

        val normalized = body.replace(Regex("<eor>", RegexOption.IGNORE_CASE), "<EOR>")
        return normalized.split("<EOR>").mapNotNull { raw ->
            val fields = parseFields(raw)
            fields.takeIf { it.isNotEmpty() }
        }
    }

    private fun parseFields(record: String): Map<String, String> {
        val fields = mutableMapOf<String, String>()
        var index = 0

        while (index < record.length) {
            val openBracket = record.indexOf('<', index)
            if (openBracket == -1) break
            val closeBracket = record.indexOf('>', openBracket)
            if (closeBracket == -1) break

            val tagContent = record.substring(openBracket + 1, closeBracket)
            val parts = tagContent.split(":", limit = 3)

            if (parts.size < 2) {
                index = closeBracket + 1
                continue
            }

            val length = parts[1].toIntOrNull()
            if (length == null) {
                index = closeBracket + 1
                continue
            }

            val fieldName = parts[0].uppercase()
            val valueStart = closeBracket + 1

            if (valueStart >= record.length) {
                index = valueStart
                continue
            }

            val valueEnd = minOf(valueStart + length, record.length)
            fields[fieldName] = record.substring(valueStart, valueEnd)
            index = valueEnd
        }

        return fields
    }

    // MARK: - Filenames

    fun activationFilename(log: Log, program: Program? = null): String = when (program) {
        Program.POTA ->
            "${log.myCallsign}@${log.potaReference ?: "POTA"}_${log.date}.adi"
        Program.SOTA -> {
            val ref = log.sotaReference
            if (ref != null) {
                "${log.myCallsign}@${ref.replace("/", "-")}_${log.date}.adi"
            } else {
                "${log.myCallsign}_SOTA_${log.date}.adi"
            }
        }
        null ->
            "${log.myCallsign}_${log.date}.adi"
    }

    fun exportAllFilename(): String {
        val formatter = SimpleDateFormat("yyyyMMdd_HHmm", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }
        return "SOTALog_${formatter.format(Date())}Z.adi"
    }

    // MARK: - QSO from fields

    fun qsoFromFields(fields: Map<String, String>, logId: Long? = null): QSO? {
        val callsign = fields["CALL"] ?: return null
        val date = fields["QSO_DATE"] ?: return null
        val timeOn = fields["TIME_ON"] ?: return null

        val band = run {
            val freqStr = fields["FREQ"]
            if (freqStr != null) {
                val freq = freqStr.toDoubleOrNull()
                if (freq != null) {
                    val derived = BandPlan.band(freq)
                    if (derived != null) return@run derived
                }
            }
            val raw = fields["BAND"]
            if (raw != null && raw in BandPlan.allBands) return@run raw
            "20m"
        }

        return QSO(
            logId = logId,
            callsign = callsign.uppercase(),
            date = date,
            timeOn = timeOn.take(4),
            frequency = fields["FREQ"]?.toDoubleOrNull(),
            band = band,
            mode = fields["MODE"] ?: "CW",
            rstSent = fields["RST_SENT"] ?: "599",
            rstReceived = fields["RST_RCVD"] ?: "599",
            name = fields["NAME"],
            qth = fields["QTH"],
            grid = fields["GRIDSQUARE"],
            sotaRef = fields["SOTA_REF"],
            potaRef = fields["SIG_INFO"],
            notes = fields["COMMENT"],
        )
    }
}
